whisper-cpp: 1.7.2 -> 1.7.5

4evy opened the pull request:
# Breaking changes:

- Switch to `whisper-*` over `whisper-cpp-*` naming convention; the reason
being: https://github.com/ggerganov/whisper.cpp/pull/2648
- The "default" binary now is `whisper-cli`

## Things done:

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

hughobrien commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 382394`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

hughobrien commented:
try removing
```
      (cmakeBool "WHISPER_BUILD_EXAMPLES" true)
```

DontEatOreo commented:
On Darwin seems to be working, so I will only remove it for Linux

hughobrien commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 382394`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

hughobrien commented:
It seems to default to true. I forced it off like this
```nix
    ++ optionals (effectiveStdenv.hostPlatform.isx86 && !effectiveStdenv.hostPlatform.isStatic) [
      (cmakeBool "GGML_BACKEND_DL" true)
      (cmakeBool "GGML_CPU_ALL_VARIANTS" true)
      (cmakeBool "WHISPER_BUILD_EXAMPLES" false)
    ]
```

But then we run afoul of assumptions about the `quantize` command
```
mv: cannot stat '/nix/store/r8fv62nkdvqilbxl18gr71haraqdhwis-whisper-cpp-1.7.4/bin/quantize': No such file or directory
```

I think this is moot anyway since the `cli` command is also an 'example'.

I've spent a silly amount of time trying to track down this issue, and my best understanding is that Nix is invoking `gcc` in a way that is confusing the complex logic in https://github.com/ggml-org/ggml/blob/master/src/ggml-cpu/CMakeLists.txt 


```
FAILED: bin/whisper-cli 
: && /nix/store/4ijy8jbsiqmj37avrk83gn2m903486mr-gcc-wrapper-14-20241116/bin/g++ -O3 -DNDEBUG -Wl,--dependency-file=examples/cli/CMakeFiles/whisper-cli.dir/link.d examples/cli/CMakeFiles/whisper-cli.dir/cli.cpp.o -o bin/whisper-cli  -Wl,-rpath,/build/source/build/src:/build/source/build/ggml/src:  examples/libcommon.a  src/libwhisper.so.1.7.4  ggml/src/libggml.so  ggml/src/libggml-base.so && :
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_get_f32_nd'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_avx'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_vsx'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_arm_fma'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_avx2'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_graph_compute_with_ctx'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_f16c'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_set_f32_nd'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_wasm_simd'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_set_f32'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_set_i32_nd'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_ssse3'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_neon'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_fp16_va'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_get_i32_nd'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_sse3'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_backend_cpu_init'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_graph_plan'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_graph_compute'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_avx512'
/nix/store/5h5ghy2qf6l91l52j6m5vx473zi38vc3-binutils-2.43.1/bin/ld: src/libwhisper.so.1.7.4: undefined reference to `ggml_cpu_has_fma'
collect2: error: ld returned 1 exit status

```

```
/data/whisper.cpp/build/examples/bench master
shell ❯ nm -D ../../ggml/src/libggml.so | grep ggml_   
                 U ggml_abort
                 U ggml_backend_dev_backend_reg
0000000000009ab0 T ggml_backend_dev_by_name
0000000000009b70 T ggml_backend_dev_by_type
0000000000009970 T ggml_backend_dev_count
00000000000099f0 T ggml_backend_dev_get
0000000000009680 T ggml_backend_device_register
                 U ggml_backend_dev_init
                 U ggml_backend_dev_name
                 U ggml_backend_dev_type
0000000000009c20 T ggml_backend_init_best
0000000000009bc0 T ggml_backend_init_by_name
0000000000009bf0 T ggml_backend_init_by_type
0000000000009c60 T ggml_backend_load
000000000000a2d0 T ggml_backend_load_all
000000000000a1f0 T ggml_backend_load_all_from_path
00000000000098b0 T ggml_backend_reg_by_name
0000000000009770 T ggml_backend_reg_count
                 U ggml_backend_reg_dev_count
                 U ggml_backend_reg_dev_get
00000000000097f0 T ggml_backend_reg_get
0000000000009530 T ggml_backend_register
                 U ggml_backend_reg_name
0000000000009e30 T ggml_backend_unload
                 U ggml_log_internal

```

`ggml` builds, but doesn't seem to expose the symbols required.  Within ggml they're simple functions like
```C
int ggml_cpu_has_ssse3(void) {
#if defined(__SSSE3__)
    return 1;
#else
    return 0;
#endif
}
```

Which I would expect to be exported, and can only conclude that the CMake complexity is somehow preventing that.

Perhaps @aviallon might have some more knowledge.

aviallon commented:
@hughobrien yes, I did try to upgrade to 1.7.4
The issue seems to becaused by the dynamically loaded backends not being built properly.
I haven't found out why exactly, but they do not build.
Removing dynamic backend loading fixes the issue, but makes whisper enormously slower.

pbsds commented on pkgs/by-name/wh/whisper-cpp/package.nix:
```diff
@@ -126,6 +123,7 @@ effectiveStdenv.mkDerivation (finalAttrs: {
       (cmakeBool "GGML_NATIVE" false)
       (cmakeBool "BUILD_SHARED_LIBS" (!effectiveStdenv.hostPlatform.isStatic))
     ]
+    ++ optional effectiveStdenv.hostPlatform.isDarwin (cmakeBool "WHISPER_BUILD_EXAMPLES" true)
```
```suggestion
    ++ [ (cmakeBool "WHISPER_BUILD_EXAMPLES" effectiveStdenv.hostPlatform.isDarwin) ]
```

4evy committed:
whisper-cpp: 1.7.2 -> 1.7.5

Diff: https://github.com/ggml-org/whisper.cpp/compare/v1.7.2...v1.7.5
Changelog: https://github.com/ggml-org/whisper.cpp/releases/tag/v1.7.5

4evy committed:
whisper-cpp: remove unused `lib.*`` funcs

4evy committed:
whisper-cpp: quote paths

DontEatOreo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 382394`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

DontEatOreo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 382394`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

DontEatOreo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 382394`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

DontEatOreo commented:
It builds on both Linux and macOS but it crashes with `57856 segmentation fault (core dumped)` when trying to transcribe :/

DontEatOreo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 382394`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

DontEatOreo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 382394`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

pbsds commented on pkgs/by-name/wh/whisper-cpp/download-models.patch:
```diff
@@ -15,39 +15,43 @@ index 1f1075b..7476c8e 100755
 -if [ "$#" -lt 1 ] || [ "$#" -gt 2 ]; then
 -    printf "Usage: %s <model> [models_path]\n" "$0"
-+if [ "$#" -ne 1 ]; then
++if [ "$#" -lt 1 ]; then
```
just a quick check: is the `-ne` -> `-lt` change intentional?
